Update Windows Crypto APIs

sbutcher-arm opened the pull request:
## Description

This is an update on PR #730 to address issues building with mingw and Visual Studio when using `cmake` generated files.

To quote the original PR.

> CryptGenRandom and lstrlenW are not permitted in Windows Store apps, meaning apps that use mbedTLS can't ship in the Windows Store. Instead, use BCryptGenRandom and wcslen, respectively, which are permitted.
> 
> Also make sure conversions between size_t, ULONG, and int are always done safely; on a 64-bit platform, these types are differnt sizes.
>

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Todos
- [X] Tests
- [] Documentation
- [X] Changelog updated
- [] Backported

gilles-peskine-arm commented on library/CMakeLists.txt:
```diff
@@ -92,7 +92,7 @@ if(CMAKE_COMPILER_IS_CLANG)
 
 if(WIN32)
-    set(libs ${libs} ws2_32)
+    set(libs ${libs} ws2_32 bcrypt)
```
This prevents building for old Windows versions, not just with Visual Studio but with e.g. mingw. Do we want that?

Not necessarily a blocker anyway, we can tell people with special needs (e.g. Windows XP) to edit the build scripts.

sbutcher-arm replied:
That's a decision we've already made. Please see the new [policy](https://tls.mbed.org/tech-updates/blog/retiring-legacy-windows-support) on it.

Basically, this is going into `devlopment` and the older LTS branches will be maintained for older Windows installations.

despair86 replied:
@gilles-peskine-arm I do maintain patchsets for recent versions of mbed TLS, as I tend to support Windows 2000/XP as a side effect of linking against the system libc `msvcrt.dll`. (Of course, depending on the rest of the project, this is not always guaranteed.)

(Yes, this has been discouraged since 1999. This is why import libraries for the recent versioned libc are included with mingw-w64. But GCC continues doing so by default because it assumes a single, systemwide `libc.so` as on UNIX and Linux)

gilles-peskine-arm commented on library/x509_crt.c:
```diff
@@ -1125,7 +1139,10 @@ int mbedtls_x509_crt_parse_path( mbedtls_x509_crt *chain, const char *path )
+        return( MBEDTLS_ERR_X509_FILE_IO_ERROR );
+
+    w_ret = MultiByteToWideChar( CP_ACP, 0, filename, lengthAsInt, szDir,
                                  MAX_PATH - 3 );
```
Preexisting: at some point someone is going to complain that `CP_ACP` is obsolescent. We should add a warning to state that this code depends on using a unibyte code page, because the output buffer size is in characters, not bytes. Changing to e.g. `CP_UTF8` would require more complex code here. See https://msdn.microsoft.com/en-us/library/windows/desktop/dd319072(v=vs.85).aspx

sbutcher-arm replied:
Given the work @irwir has been doing on PR #1427, I think I'd rather this was raised as a bug, rather than a comment in the code.

If you agree, could you please raise the bug?

gilles-peskine-arm replied:
I don't see any functional bug here, just a maintenance issue that could lead to a bug in the future. This needs to be a comment in the code. What would be the bug? To switch to `CP_UTF8`? I don't know if it's the right thing. For all I know, `CP_ACP` is fine because we want to treat the file name as just some string to be passed (semantically) unchanged and we don't care about encoding.

gilles-peskine-arm replied:
It seems to me that #1427 would bypass this issue. I'd still rather have the comment, even in the backports.

sbutcher-arm replied:
I think I should have said raise it as an issue, not a bug, and that implies a fault.

I'm not sure my point was obvious enough to I'll explain it again. @irwir has created a PR to change the strings that touch the platform from `char`'s to `TCHAR`'s to allow portability between ASCII/UTF-8 and Unicode, including changing interface code so the public interface on Windows can accept `TCHAR` strings.

I do think we need to consider how we interface with Windows and what we should do about this, and consider if and how we address the issue.

My suggestion of raising an issue was really about that.

Anyway - I'll take your point and add a comment.

kevinmkane replied:
I didn't want to change this code more than I had to when I originally added this, but since you've brought it up, it looks like the whole point of these conversions is to call FindFirstFileW and FindNextFileW, and then to convert their results back to narrow character strings to be used.

Another possibility is to dispense with wide character strings entirely, since the mbedTLS API always takes narrow character strings, and use the narrow character string Windows equivalents: the FindFirstFileA and FindNextFileA functions and the WIN32_FIND_DATAA type. Then the platform will do all the string conversions internally and this code can only deal in narrow character strings.

If you decide to move to a TCHAR API approach then all of this is moot, of course, but for however long you stick with narrow character strings, you might consider this.

irwir replied:
> this code depends on using a unibyte code page

Is 'unibyte' word about one byte being one character?
In that case, ANSI is no more unibyte than UTF-8. Just remember ANSI code pages for Oriental languages.
Whereever needed, Windows provide two APIs for the same function: one ANSI, and one Unicode.
No conversion should be needed when strings and API were both of ANSI kind.

gilles-peskine-arm replied:
Yes, “unibyte” means that 1 character = 1 byte. Thanks @irwir, I didn't know that ANSI could be multibyte (I'm not familiar with Windows APIs at all). Given that `szDir` is a `WCHAR` array, is the size correct? In other words, does an ANSI character always fit in a `WCHAR`?

irwir replied:
> does an ANSI character always fit in a WCHAR?

That could be expected in most cases, but not guaranteed because `WCHAR` implies Unicode with its peculiarities.
`MAX_PATH` dedermines the number of array elements, not the number of characters (see description of  `WIN32_FIND_DATA` structure). And, as said, conversion to wide character is not needed.

irwir replied:
> Another possibility is to dispense with wide character strings entirely, since the mbedTLS API always takes narrow character strings

The cryptographic part actually uses bytes and does not care about characters.
Characters appear when API refers to path/file names. As I was told, it was supposed to be  in UTF-8 encoding, and that is not what 'narrow' strings were supposed to mean in Windows code.

gilles-peskine-arm commented on library/Makefile:
```diff
@@ -155,7 +155,7 @@ libmbedcrypto.dylib: $(OBJS_CRYPTO)
 libmbedcrypto.dll: $(OBJS_CRYPTO)
 	echo "  LD    $@"
-	$(CC) -shared -Wl,-soname,$@ -Wl,--out-implib,$@.a -o $@ $(OBJS_CRYPTO) -lws2_32 -lwinmm -lgdi32 -static-libgcc $(LOCAL_LDFLAGS) $(LDFLAGS)
+	$(CC) -shared -Wl,-soname,$@ -Wl,--out-implib,$@.a -o $@ $(OBJS_CRYPTO) -lws2_32 -lbcrypt -lwinmm -lgdi32 -static-libgcc $(LOCAL_LDFLAGS) $(LDFLAGS)
```
Blocker: you need to update something for the non-shared build as well. Under Ubuntu 17.10:
```
$ make CC=i686-w64-mingw32-gcc AR=i686-w64-mingw32-ar LD=i686-w64-minggw32-ld CFLAGS='-Werror -Wall -Wextra' WINDOWS_BUILD=1
…
  CC    test_suite_entropy.c
../library/libmbedcrypto.a(entropy_poll.o):entropy_poll.c:(.text+0x78): undefined reference to `BCryptGenRandom@16'
collect2: error: ld returned 1 exit status
Makefile:309: recipe for target 'test_suite_entropy.exe' failed
```

sbutcher-arm replied:
Fixed in 63450c3.

gilles-peskine-arm commented:
We should document the minimum mingw version in README. If not, at least on the website.

sbutcher-arm commented:
@gilles-peskine-arm - We already have an open task in JIRA (IOTSSL-1042) to update the mingw documentation.

sbutcher-arm commented:
Review comments from @gilles-peskine-arm addressed. Please review my responses and respond or approve.

Thanks!

sbutcher-arm commented:
Updated with a comment on the codepage`CP_ACP`. 

@gilles-peskine-arm / @Patater - please re-review and approve.

Patater commented:
Testing done (in addition to CI checks):
Built benchmark.exe with Makefile and mingw. Ran benchmark.exe on wine; it worked.

Patater commented:
The provided solution files for x64 builds fail to build. This is either a regression introduced by this PR or because the PR is based on old versions of those files. If the latter, please consider rebasing on a more recent point in history to get our recent fixes to those solution files.

sbutcher-arm commented:
@Patater - You're right. It was a regression introduced by the PR, which is now fixed. The PR is [now passing the Windows CI](https://jenkins-internal.mbed.com/job/mbedtls-release/282/console) and builds for Win32 and x64 with Visual Studio 2015 on my development machine.

@gilles-peskine-arm / @Patater - please review and approve as you see fit.

Please note I've also removed the obsolete Visual Studio 6 project files.

gilles-peskine-arm commented:
Thanks @Patater . Rather than change an old commit, which will make the history inaccurate and makes it harder to diagnose problems later, I'd rather have a fix-up commit that says ”do &lt;such-and-such&gt;. This reverts the incorrect change made in commit 1234 which caused &lt;problem&gt;.”

sbutcher-arm commented:
The problems in b862f4c were introduced when I rebased the original PR against the current `development` branch, and I inadvertently introduced `mbedtls.lib` into the project files which reverted a change introduced in PR #1384.

Rather than layer on more commits to fix the original botched rebase, I rebased the original PR again, and have added back the other commits required to address the other issues and review feedback again.

The PR is now passing the Windows CI with the exception of the known issues and intermittent problems, and also issue #1430, which is causing some Release builds to fail.

Given I've rebased the PR again, this PR needs a review from beginning to end again.

dgreen-arm commented:
LGTM, except there seems to be a minor inconsistency with the casing of the variable names: len_as_ulong and lengthAsInt.

mazimkhan commented:
This PR now passes Windows cross build on Linux after updating CI slave to newer version.
https://jenkins-internal.mbed.com/job/mbedtls-release/358/

sbutcher-arm commented:
I've rebased this PR, and kicked off a CI job on buildbot, and a release job on Jenkins.

@Patater / @gilles-peskine-arm - Could you please re-review and approve this PR? Or if you're unavailable, please let me know so I can find substitutes. Thanks.

gilles-peskine-arm commented:
@sbutcher-arm Given my limited bandwidth and my lack of familiarity with Windows, please find someone else for this PR.

Patater commented on scripts/data_files/vs2010-main-template.vcxproj:
```diff
@@ -107,6 +110,12 @@
+      <ShowProgress>NotSet</ShowProgress>
+      <AdditionalDependencies>bcrypt.lib;kernel32.lib;user32.lib;gdi32.lib;winspool.lib;comdlg32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;odbc32.lib;odbccp32.lib;%(AdditionalDependencies)</AdditionalDependencies>
+      <AdditionalLibraryDirectories>Debug</AdditionalLibraryDirectories>
+>>>>>>> Replace Windows APIs that are banned in Windows Store apps:visualc/VS2010/md5sum.vcxproj
```
Probably not a good idea to have conflict markers here...

sbutcher-arm replied:
Ooops! Sorry. Now fixed.

sbutcher-arm commented:
retest

sbutcher-arm commented:
Failed Travis CI due to stale Visual Studio project files, found by `tests/scripts/check-generated-files.sh`. I'll refresh them and update the PR.

sbutcher-arm commented:
Fixed the stale project files and the merge error. Everything should now pass the CI.

Patater commented on ChangeLog:
```diff
@@ -247,6 +247,8 @@ Changes
    * Use (void) when defining functions with no parameters. Contributed by
      Joris Aerts. #678
+   * Update Win32 APIs used to remove those not permitted by the Windows Store.
+     Fix provided by Kevin Kane, Microsoft. #635
```
Git has expertly added this as a "Changes" entry to "mbed TLS 2.8.0". This needs fixing.

sbutcher-arm replied:
I'll fix that in gatekeeping. It's not worth re-running the CI to fix it in the PR.

gilles-peskine-arm replied:
@Patater Git doesn't know how to merge changelog entries (not without [help](https://github.com/ARMmbed/mbedtls/pull/1195) anyway) so it's routine for us to correct the location of added entries when merging.

Patater replied:
Note to gatekeeper: Please fix this.

This can easily go unnoticed in gatekeeping as it doesn't cause a conflict. I'd recommend making the change in the PR, but if that can't be done for some reason, then please keep this fix in mind when gatekeeping.

Patater commented:
Release testing job is running at https://jenkins-internal.mbed.com/view/mbed-tls/job/mbedtls-release/508/

k-stachowiak commented on library/x509_crt.c:
```diff
@@ -1126,6 +1139,7 @@ int mbedtls_x509_crt_parse_path( mbedtls_x509_crt *chain, const char *path )
     char filename[MAX_PATH];
     char *p;
     size_t len = strlen( path );
+    int lengthAsInt = 0;
```
I agree with Darryl's comment - this name doesn't follow the Mbed TLS conding standard.

sbutcher-arm replied:
Fixed.

sbutcher-arm commented:
@k-stachowiak / @Patater / @gilles-peskine-arm - I've fixed the style issue @k-stachowiak identified. Could you please re-review?

Scheduled a Windows release job as build #552.

sbutcher-arm commented:
Sadly, the PR failed `all.sh` release job:

Extract from the log:
```
[all_sh] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[all_sh] FAILED: 5
[all_sh] build: create and build yotta module: tests/scripts/yotta-build.sh -> 1
[all_sh] build: Windows cross build - mingw64, make (Link Library): command make CC=i686-w64-mingw32-gcc AR=i686-w64-mingw32-ar LD=i686-w64-minggw32-ld CFLAGS=-Werror -Wall -Wextra WINDOWS_BUILD=1 lib programs -> 2
[all_sh] build: Windows cross build - mingw64, make (Link Library): command make CC=i686-w64-mingw32-gcc AR=i686-w64-mingw32-ar LD=i686-w64-minggw32-ld CFLAGS=-Werror WINDOWS_BUILD=1 tests -> 2
[all_sh] build: Windows cross build - mingw64, make (DLL): command make CC=i686-w64-mingw32-gcc AR=i686-w64-mingw32-ar LD=i686-w64-minggw32-ld CFLAGS=-Werror -Wall -Wextra WINDOWS_BUILD=1 SHARED=1 lib programs -> 2
[all_sh] build: Windows cross build - mingw64, make (DLL): command make CC=i686-w64-mingw32-gcc AR=i686-w64-mingw32-ar LD=i686-w64-minggw32-ld CFLAGS=-Werror -Wall -Wextra WINDOWS_BUILD=1 SHARED=1 tests -> 2
[all_sh] !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

gilles-peskine-arm commented:
@sbutcher-arm Maybe the new API isn't supported by the version of Mingw on the CI? IIRC at some point in our Windows crypto API saga there were proposals that didn't build with the toolchain version in Ubuntu 16.04 but did build with the toolchain version in Ubuntu 18.04.

sbutcher-arm commented:
I think you're right @gilles-peskine-arm, and the issue is with the mingw version on the Jenkins CI. We're trying to get hold of the dockerfile and update it.

sbutcher-arm commented:
This PR passes with the correct mingw version, so I'm going accept the PR as-is, and fix the Jenkins CI as quickly as I can independently. In the meantime, all.sh will continue to fail on Jenkins CI until we fix it, for the mingw tests.

cc: @dgreen-arm, @Patater

dgreen-arm commented:
This has failed on Visual Studio builds for Release and x64. Looking at the PR, it looks like the configuration for Release|x64 wasn't updated like the others were.

sbutcher-arm commented:
Demoting to 'needs work' until we fix this build problem.

sbutcher-arm commented:
Now passing on Jenkins CI for 'Release|x64'.

@Patater / @k-stachowiak - Can you please review the last commit, and approve, so I can merge this? Thanks.

Patater commented on visualc/VS2010/gen_random_havege.vcxproj:
```diff
@@ -165,7 +165,7 @@
       <OptimizeReferences>true</OptimizeReferences>
       <AdditionalLibraryDirectories>Release</AdditionalLibraryDirectories>
-      <AdditionalDependencies>%(AdditionalDependencies);</AdditionalDependencies>
+      <AdditionalDependencies>bcrypt.lib;kernel32.lib;user32.lib;gdi32.lib;winspool.lib;comdlg32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;odbc32.lib;odbccp32.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
Should we leave this variable unexpanded? Here and in other similar places.

sbutcher-arm replied:
What I've done in the most recent commit is to copy the change that was present in the other passing configurations such as 'Release|Win32' into 'Release|x64' which was failing. In a way - I can't say I fully understand the configuration file itself, and have just copied what works elsewhere to what doesn't work.

Given we don't fully understand it, and we generate these files, so have to adjust a template not the files themselves, I'd prefer to be conservative and change only what we have to.

despair86 replied:
that would work, it may have been an oversight as the IDE normally fills in that particular list of import archives

sbutcher-arm commented:
Passes Jenkins CI in build #871.

despair86 commented:
that would work, may have been an oversight as the IDE normally fills in that particular list of import archives

-R

sbutcher-arm commented:
This PR failed in my attempts to merge it causing the CI release job to fail. I've rebased the PR (will need re-approval - sorry), and set off a new CI [release job #894](https://jenkins-internal.mbed.com/view/mbed-tls/job/mbedtls-release/894/).

sbutcher-arm commented:
I've tested with the Docker image in the CI, and the issue is the version of mingw is too old, being `i686-w64-mingw32-gcc (GCC) 5.3.1 20160211`.

This is too old, and even Ubuntu 17.10 currently has a package version of `i686-w64-mingw32-gcc (GCC) 6.3.0 20170415`.

This issue is with the CI and needs to be updated to merge this PR.

sbutcher-arm commented:
retest

joycepg commented:
i'm sad this PR hasn't been merged yet, but since it hasn't, i thought i'd point out an alternative that causes less changed files. 
the change to bcrypt api in entropy_poll.c required many changes in the makefiles, project files, cmake files, etc. instead of all these explicit reference to link against bcrypt.lib, adding this one line:

`#pragma comment(lib,"bcrypt.lib")`

into entropy_poll.c (eg just after `#include <bcrypt.h>`) causes the dependency on bcrypt.lib to be noted inside entropy_poll.**obj**, and therefore it is pulled in by the linker automatically.

ackh commented:
Is there any chance that this change ends up in a release any time soon? What is blocking it? Can I do something to help out?

sbutcher-arm commented:
I'll pick this PR up again, rebase it and resubmit it. Hopefully we can get it in this time.

despair86 commented:
@joycepg that trick only works for Microsoft C.

mpg commented:
Note: there was a PR on the crypto side that may or may not contain changes that need to be picked up here now: https://github.com/ARMmbed/mbed-crypto/pull/341

Zopolis4 commented:
Any news on this? I'm trying to build a project that uses mbedtls for UWP.

akien-mga commented:
For the record, we've been using that PR downstream in Godot Engine for a few years for our UWP build. Here's a diff of this PR compatible with mbedTLS 2.28.1: https://github.com/godotengine/godot/blob/master/thirdparty/mbedtls/patches/1453.diff

Our UWP port doesn't see a lot of use though so I can't attest for how well this solves the issue, and whether mbedTLS works properly on UWP.

akien-mga commented:
> CryptGenRandom and lstrlenW are not permitted in Windows Store apps, meaning apps that use mbedTLS can't ship in the Windows Store. Instead, use BCryptGenRandom and wcslen, respectively, which are permitted.

The `lstrlenW` part seems to have been resolved by #7081 (in 3.4.0) and #7086 (in 2.28.3).

Not sure what's the state with `CryptGenRandom` on UWP nowadays, for now I'll keep that part of the patch but I'll see if it can be dropped too.

robUx4 commented:
For the record, `CryptGenRandom` is allowed in UWP since Windows 19H1.

akien-mga commented:
Superseded by https://github.com/Mbed-TLS/mbedtls/pull/8047 I believe?
